functioncall/gpact/src/main/solidity/CrosschainControl.sol

```diff
@@ -716,45 +714,48 @@ contract CrosschainControl is
                 (uint256, bytes32, uint256[], address[], bool, bytes)
             );
 
-            // Recall Segment event is defined as:
-            // event Segment(uint256 _crossBlockchainTransactionId, bytes32 _hashOfCallGraph, uint256[] _callPath,
-            //        address[] _lockedContracts, bool _success, bytes _returnValue);
-            //            uint256 crossBlockchainTxId = BytesUtil.bytesToUint256(segmentEvent, 0);
-            //            bytes32 hashOfCallGraphFromSegment = BytesUtil.bytesToBytes32(segmentEvent, 0x20);
-            //            uint256 locationOfCallPath = BytesUtil.bytesToUint256(segmentEvent, 0x40);
-            // Not needed: uint256 locationOfLockedContracts = BytesUtil.bytesToUint256(segmentEvent, 0x60);
-            //            uint256 success = BytesUtil.bytesToUint256(segmentEvent, 0x80);
-            //            uint256 locationOfReturnValue = BytesUtil.bytesToUint256(segmentEvent, 0xA0);
-            //            uint256 lenOfReturnValue = BytesUtil.bytesToUint256(segmentEvent, locationOfReturnValue);
-            //            bytes memory returnValue = BytesUtil.slice(segmentEvent, locationOfReturnValue + 0x20, lenOfReturnValue);
-            //            uint256 lenOfCallPath = BytesUtil.bytesToUint256(segmentEvent, locationOfCallPath);
-
             require(
                 crossBlockchainTxId == _crosschainTxId,
                 "Transaction id from segment and root do not match"
             );
             require(
-                hashOfCallGraph == hashOfCallGraphFromSegment,
+                _hashOfCallGraph == hashOfCallGraphFromSegment,
                 "Call graph from segment and root do not match"
             );
 
-            // Segments with offset zero of Root calls are the only ones that can call segments.
-            require(callPath[callPath.length - 1] == 0);
-            //            require(lenOfCallPath == callPath.length || lenOfCallPath == callPath.length+1, "Bad call path length for segment");
+            // The caller must be a segment that is calling other segments or the
+            // root call (which also is calling segments). These both have the call path
+            // for the caller will have 0 as the offset in the call path array for the
+            // previous level in the call path.
+            require(_callPath[_callPath.length - 1] == 0, "Invalid caller");
```

Comment:
I might be missing something but is there a reason this `require` is in the for loop? `_callPath` refers to the function arg and does not seem to change during the loop iterations?

Reply:
do we have tests that handle those requires? not the happy path

Reply:
@ermyas , good point. This check should happen outside the loop.

Reply:
@ricardolyn, we need to do a comprehensive set of tests for the code base. This is an area where these is only happy path testing.

Reply:
but why if we are adding this new checks, we cannot add the tests at the same time? doing it after will create a higher probability of getting lost and missing edge cases.